R/get_vpts_aloft.R

```diff
@@ -0,0 +1,122 @@
+#' @param coverage A data.frame containing the coverage of the aloft data
+#'   repository. If not provided, it will be fetched from via the internet.
+#' @return A list of vpts data.frames
+#'
+#' @importFrom dplyr .data
+#' @importFrom lubridate %within%
+#'
+#' @examples
+#' get_vpts_aloft(
+#'   "bejab",
+#'   lubridate::interval("20240305", "20240307"),
+#'   "baltrad"
+#' )
+#'
+#' @noRd
+get_vpts_aloft <- function(radar_odim_code,
+                           rounded_interval,
+                           source,
+                           coverage = aloft_data_coverage()) {
+  # rename source argument for readability
+  selected_source <- source
+
+  # Check that only one radar is provided (string of length 1)
+  if (!rlang::is_string(radar_odim_code)) {
+    cli::cli_abort(
+      "Please provide (only one) radar as a character vector of length 1.",
+      class = "getRad_error_radar_not_single_string"
+    )
+  }
+
+  # Check that radar_odim_code is a single 5 character string
+  if (nchar(radar_odim_code) != 5) {
+    cli::cli_abort(
+      "Radar ODIM code must be a single 5 character string.",
+      class = "getRad_error_radar_odim_code_invalid"
+    )
+  }
+
+  # Check if the requested radars are present in the coverage
+  if (!all(radar_odim_code %in% coverage$radar)) {
+    cli::cli_abort(
+      "Radar(s) not found in ALOFT coverage:
+      {radar_odim_code[!radar_odim_code %in% coverage$radar]}.",
+      class = "getRad_error_aloft_radar_not_found"
+    )
+  }
+
+  # Check if the requested date radar combination is present in the coverage
+  filtered_coverage <- dplyr::filter(
+    coverage,
+    .data$radar %in% radar_odim_code,
+    .data$date %within% rounded_interval,
+    .data$source == selected_source
+  )
+  at_least_one_radar_date_combination_exists <- nrow(filtered_coverage) > 0
+
+  if (!at_least_one_radar_date_combination_exists) {
+    cli::cli_abort(
+      "No data found for the requested radar(s) and date(s).",
+      class = "getRad_error_date_not_found"
+    )
+  }
+
+  # Filter the coverage data to the selected radars and time interval and
+  # convert into paths on the aloft s3 storage
+  ## We need to use the rounded interval because coverage only has daily
+  ## resolution
+  s3_paths <- filtered_coverage |>
+    dplyr::pull(.data$directory) |>
+    # Replace hdf5 with daily to fetch vpts files instead of hdf5 files
+    string_replace("hdf5", "daily") |>
+    # Construct the filename using glue mapping over every path.
+    purrr::map_chr(\(path){
+      glue::glue(
+        "{dir}/{radar}_vpts_{year}{month}{day}.csv",
+        dir = string_extract(path, ".+/.+/.+/[0-9]{4}"),
+        radar = string_extract(path, "(?<=daily/).{5}"),
+        year = string_extract(path, "[0-9]{4}"),
+        month = string_extract(path, "(?<=/)[0-9]{2}(?=/)"),
+        day = string_extract(path, "[0-9]{2}$")
+      )
+    })
+
+  # Read the vpts csv files
+  aloft_data_url <- "https://aloftdata.s3-eu-west-1.amazonaws.com"
```

Comment:
Should this be an option to be able to point to other repositories?

Reply:
Get from option, mention in function details.